e2e/nomostest/testshell/helm_client.go

```diff
@@ -0,0 +1,123 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testshell
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"kpt.dev/configsync/e2e"
+)
+
+// HelmClient provides helm and crane clients for connecting to an Helm repository.
+type HelmClient struct {
+	// DockerConfig is the value to use for DOCKER_CONFIG.
+	// Crane uses DOCKER_CONFIG for auth settings.
+	// Helm uses DOCKER_CONFIG for default auth settings, when not specified in
+	// HELM_REGISTRY_CONFIG.
+	DockerConfig string
+
+	// ConfigHome is the value to use for HELM_CONFIG_HOME.
+	//
+	// Defaults:
+	// - Linux: ${HOME}/.config/helm
+	// - Mac: ${HOME}/Library/Preferences/helm
+	//   https://github.com/helm/helm/blob/main/pkg/helmpath/lazypath_darwin.go#L28
+	// - Windows: ${APPDATA}/helm
+	//   https://github.com/helm/helm/blob/main/pkg/helmpath/lazypath_windows.go#L22
+	//
+	// Affects default values for HELM_REGISTRY_CONFIG, HELM_REPOSITORY_CONFIG
+	// HELM_REPOSITORY_CACHE, and HELM_CACHE_HOME.
+	ConfigHome string
+
+	// Shell used to execute helm and crane client commands.
+	Shell *TestShell
+}
+
+// NewHelmClient constructs a new HelmClient that stores config in
+// sub-directories under the specified `parentPath`.
+func NewHelmClient(parentPath string, shell *TestShell) *HelmClient {
+	cfg := &HelmClient{
+		DockerConfig: filepath.Join(parentPath, "helm-docker"),
+		ConfigHome:   filepath.Join(parentPath, "helm"),
+		Shell: &TestShell{
+			Context: shell.Context,
+			Env:     append([]string{}, shell.Env...), // Copy parent shell environment
+			Logger:  shell.Logger,
+		},
+	}
+	// Add environment vars to be used for all commands with this shell.
+	// Because the shell is a copy, it won't affect the parent shell.
+	cfg.Shell.Env = append(cfg.Shell.Env,
+		fmt.Sprintf("DOCKER_CONFIG=%s", cfg.DockerConfig),
+		fmt.Sprintf("HELM_CONFIG_HOME=%s", cfg.ConfigHome))
+
+	shell.Logger.Infof("DOCKER_CONFIG for helm will be created under %s", cfg.DockerConfig)
+	shell.Logger.Infof("HELM_CONFIG_HOME will be created under %s", cfg.ConfigHome)
+	shell.Logger.Infof("Connect to the %q Helm registry using:\nexport DOCKER_CONFIG=%s\nexport HELM_CONFIG_HOME=%s",
+		*e2e.HelmProvider, cfg.DockerConfig, cfg.ConfigHome)
+
+	return cfg
+}
+
+// Login uses gcloud to configure docker auth credentials used by helm and crane.
+//
+// Also explicitly calls `helm registry logout` and `crane auth logout` to
+// remove any saved tokens for the specified registry, as these take precedence.
+//
+// Files modified:
+// - `${DOCKER_CONFIG}/config.json` (crane/docker)
+// - `${HELM_CONFIG_HOME}/registry/config.json` (helm)
+func (hc *HelmClient) Login(registryHost string) error {
+	if out, err := hc.Helm("registry", "logout", registryHost); err != nil {
```

Comment:
Is the explicit logout still necessary now that a config path is being set in the shell's environment?

Reply:
As long as the cleanup happens and the environment isn't re-used while dirty, then yeah, we shouldn't need to log out before logging it. 

Is that happening? 

I think the TmpDir doesn't actually get nuked when a test fails when running --debug. If you don't manually delete it, will it use the same one next time based on the name of the environment/cluster? Or is it randomized?

Reply:
removed